### ISSUE TITLE
WordPress != WordPress-Core,WordPress-Docs,WordPress-Extra

### ISSUE THREAD

**Issue by GaryJones** [WordPress-Coding-Standards/WordPress-Coding-Standards#1440]
After refreshing https://github.com/GaryJones/pwa-wp/tree/phpcs-refresh, no violations are reported when running `phpcs`. All good.

If I change:

```xml
<rule ref="WordPress-Core"/>
<rule ref="WordPress-Docs"/>
<rule ref="WordPress-Extra"/>
```

to:

```xml
<rule ref="WordPress"/>
```

Then new violations are reported:

```
FILE: wp-includes/class-wp-service-workers.php
-----------------------------------------------------------------------------------------------------------------------------------------
FOUND 2 ERRORS AFFECTING 1 LINE
-----------------------------------------------------------------------------------------------------------------------------------------
 117 | ERROR | Missing wp_unslash() before sanitization. (WordPress.Security.ValidatedSanitizedInput.MissingUnslash)
 117 | ERROR | Detected usage of a non-sanitized input variable: $_SERVER
     |       | (WordPress.Security.ValidatedSanitizedInput.InputNotSanitized)
-----------------------------------------------------------------------------------------------------------------------------------------
```

Is that supposed to happen?

cc @westonruter

**Comment by jrfnl** [comment]
This is a known non-issue.

When we moved/renamed the original badge of sniffs as discussed in #1157, most were added to the `Extra` ruleset in #1261, save three for which more discussion/bug fixes were needed. The sniff you refer to is one of those three.

This was annotated in the PR at the time.

> Note: There are three more candidates for this:
> * DB.DirectDatabaseQuery
> * DB.SlowDBQuery
> * Security.ValidatedSanitizedInput
> 
> I've not included the first two (yet) as those may need some more discussion first.
> The third one is not included as it throws a lot of noise messages and has quite some open issues surrounding it, which should be addressed first IMO.

See #1261

**Comment by GaryJones** [comment]
Sorry @jrfnl, I don't follow.

Why does this violation appear when checking against `WordPress`, but not when checking against the three sub-standards together?

**Comment by jrfnl** [comment]
> Why does this violation appear when checking against WordPress, but not when checking against the three sub-standards together?

Because the above mentioned three sniffs are not included in any of the three sub-standards.

The `WordPress` ruleset included the three sub-standards + every single sniff in the `Sniffs` directory, which is why these three will be included when you run against the `WordPress` standard.

The "_every single sniff in the `Sniffs` directory_" bit is default behaviour of PHPCS and is part of the different between a _custom ruleset_ and a _standard_.
* A standard can have a `Sniffs` sub-directory and any sniffs in that sub-directory will be _automatically_ registered for the standard.
* A custom ruleset can include sniffs and configuration, but cannot _add_ sniffs.
    The `WordPress-...` rulesets are basically custom rulesets (no `Sniffs` sub-directory), but as they are shipped within the WPCS package, they are registered as standards.

This is the same principle of why we were having so many issues with the deprecation warnings being thrown, even though the VIP standard was not being included.

**Comment by GaryJones** [comment]
So, at least while those three sniffs need discussion and are not included in `Extra`, it's best to stick with:

```xml
<rule ref="WordPress-Core"/>
<rule ref="WordPress-Docs"/>
<rule ref="WordPress-Extra"/>
```

in a project's ruleset?

**Comment by jrfnl** [comment]
@GaryJones That depends on your intention with the custom ruleset.

The simple answer is that with regards to these three sniffs, the behaviour of the `WordPress` ruleset and the `WordPress-Extra` ruleset has not changed in 1.0.0.
The sniffs were previously already included in the `WordPress` ruleset and have never been included in the `WordPress-Extra` ruleset. I.e.: no change.

So if you want the "same" behaviour as before (aside from all the other changes in 1.0.0), leave the ruleset as is, whether that is `WordPress` or a combination of sub-standards.

**Comment by GaryJones** [comment]
Understood, thank you.

Worth adding something to the readme, regarding explaining the above?

**Comment by jrfnl** [comment]
@GaryJones Did you still want to add a note about this to the Readme ? If so, please do. Otherwise, let's close this.

**Comment by GaryJones** [comment]
Is there any way of easily identifying which sniffs from the `Sniffs` directory are NOT part of one of the three rulesets?

**Comment by jrfnl** [comment]
@GaryJones Well, there's the list above. Other than that, I think it would be a case of doing a (manual) comparison between the output of these commands:
```bash
phpcs -e --standard=WordPress-Extra
phpcs -e --standard=WordPress
```

... where the focus would be on what's listed under the "WordPress" heading.

I mean, we know that there are no WPCS native sniffs in the `Docs` ruleset, so we don't need to take it into account.
And the `WordPress-Extra` ruleset included `WordPress-Core`, so comparing the output of the above two commands should give a conclusive answer.

I've just done so and the outcome confirms what we already knew:
```diff
  WordPress.DB.DirectDatabaseQuery
  WordPress.DB.SlowDBQuery
  WordPress.Security.ValidatedSanitizedInput
  WordPress.Utils.I18nTextDomainFixer
```

Note: `WordPress.Utils.I18nTextDomainFixer` is a sniff which shouldn't normally run anyway as it is a utility to switch textdomains, so that being there is correct and expected.

**Comment by GaryJones** [comment]
Are you intending to include the top three within a standard before 3.0.0?

**Comment by jrfnl** [comment]
There are included in the `WordPress` standard and AFAIC, that's as it should be as per https://github.com/WordPress/WordPress-Coding-Standards/issues/1440#issuecomment-408690186.

The decision not to include them in `WordPress-Extra` was previously taken to prevent a BC-break, even though the related changes were made at a major release.

I'm open to hearing a case being made for changing that now, but not something I'm terribly concerned about.

**Comment by leewillis77** [comment]
Apologies for the cross-posting on the other issue - it had skipped my attention that this issue was still open. I've opened a PR that adds an explanation to the README to clarify that WordPress includes additional checks that aren't runb by any of the subsets which is certainly a help. 

Longer term I think having `WordPress` include only subsets and no direct sniffs itself would be the more useful outcome for a number of reasons:

Principally, it's the more obvious interpretation and thereby lowers the risk that someone will misunderstand what checks are being run. I'm a big fan of the principle of least-surprise.

It's more flexible. At the minute, the subsets are only really useful as "exclusions", not "inclusions" since you can't get some tests other by running `WordPress` and then disabling the ones you don't want. I initially was going to also note in the README changes that using the subsets as "exclusions" was the preferred approach but after a bit of digging that doesn't really work either since they can't be excluded on the command line, e.g.

Running with this in a `phpcs.xml` works:
```xml
<?xml version="1.0"?>
<ruleset>
        <rule ref="WordPress">
                <exclude name="WordPress-Extra"/>
        </rule>
        <file>test.php</file>
</ruleset>
```

However, the same effect can't be achieved on the command line:
```bash
$ phpcs --standard=WordPress --exclude=WordPress-Extra
ERROR: The --exclude option only supports sniff codes.
Sniff codes are in the form "Standard.Category.Sniff".

The following problems were detected:
* Standard codes are not supported: WordPress-Extra
```

I'd love to see all of the sniffs in `WordPress` be moved into one of the existing (or a new) subset at the earliest opportunity (4.x I guess?), and then the subsets can be used additively to build up the required level of checks.

**Comment by jrfnl** [comment]
> Longer term I think having WordPress include only subsets and no direct sniffs itself would be the more useful outcome

Technically, that's an impossibility as any standard containing sniffs will automatically include every single sniff in that standard.

In practice, though, I believe we can make this work by making sure that all `WordPress` sniffs are included in at least one of the sub-rulesets.

Either way, to get to a place where this would be possible, I think it would be prudent to check whether the previous objections against the sniffs being moved into `Extra` still apply.
That means someone will need to check open issues related to these sniffs etc.

If the objections no longer apply, we don't need to wait until 4.0, any minor will do as it is pretty common for new rules to be added to rulesets in a minor.

I do wonder about including the `WordPress.Utils.I18nTextDomainFixer` sniff in a sub-ruleset though as that sniff is - and always was - intended as a tool to run once and running it for every PHPCS run makes no sense whatsoever.

For the time being, I'd like to suggest for the PR (#2661) to be put on HOLD until we've had a proper discussion about a path forward.


> At the minute, the subsets are only really useful as "exclusions", not "inclusions" ...

What I read from here on down is a lot of misinformation or maybe lack of understanding ? As this ticket is not about teaching how rulesets work, I won't go into this as I don't want to distract from the real topic being discussed, but please know that the above doesn't make much, if any, sense to me.
